src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -0,0 +1,139 @@
+            var reporter = crawlerService.GetProgressReporter(workspace);
+
+            Started(reporter.InProgress);
+
+            reporter.ProgressChanged += OnSolutionCrawlerProgressChanged;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+
+            _options = new TaskHandlerOptions()
+            {
+                Title = ServicesVSResources.Live_code_analysis,
+                ActionsAfterCompletion = CompletionActions.None
+            };
+        }
+
+        private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+        {
+            // there is no concurrent call to this method since IDiagnosticService will serialize all
+            // events to preserve event ordering
+            var solution = e.Solution;
+            if (solution == null)
+            {
+                return;
+            }
+
+            var document = solution.GetDocument(e.DocumentId);
+            if (document != null)
+            {
+                ChangeProgress(string.Format(ServicesVSResources.Analyzing_0, document.Name ?? document.FilePath ?? "..."));
+                return;
+            }
+
+            var project = solution.GetProject(e.ProjectId);
+            if (project != null)
+            {
+                ChangeProgress(string.Format(ServicesVSResources.Analyzing_0, project.Name ?? project.FilePath ?? "..."));
+                return;
+            }
+        }
+
+        private void OnSolutionCrawlerProgressChanged(object sender, bool running)
+        {
+            // there is no concurrent call to this method since ISolutionCrawlerProgressReporter will serialize all
+            // events to preserve event ordering
+            Started(running);
+        }
+
+        private void Started(bool running)
+        {
+            if (running)
+            {
+                // if there is any pending one. make sure it is finished.
+                _currentTask?.TrySetResult(true);
+
+                var taskHandler = _taskCenterService.PreRegister(_options, data: default);
+
+                _currentTask = new TaskCompletionSource<bool>();
+                taskHandler.RegisterTask(_currentTask.Task);
+
+                var data = new TaskProgressData
+                {
+                    ProgressText = null,
+                    CanBeCanceled = false,
+                    PercentComplete = null,
+                };
+
+                // report initial progress
+                taskHandler.Progress.Report(data);
+
+                // set handler
+                _taskHandler = taskHandler;
+            }
+            else
+            {
+                // stop progress
+                _currentTask?.TrySetResult(false);
+                _currentTask = null;
+
+                _taskHandler = null;
+            }
+        }
+
+        private void ChangeProgress(string message)
+        {
+            if (message == _lastMessage)
+            {
+                return;
```

Comment:
:bulb: If this is frequently hit, a better fast path would avoid the calls to `string.Format`.

Reply:
good idea